File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/SoundManager.java
package com.habitrpg.android.habitica.helpers;

import com.habitrpg.android.habitica.HabiticaBaseApplication;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.schedulers.Schedulers;

public class SoundManager {
    public static String SoundAchievementUnlocked = "Achievement_Unlocked";
    public static String SoundChat = "Chat";
    public static String SoundDaily = "Daily";
    public static String SoundDeath = "Death";
    public static String SoundItemDrop = "Item_Drop";
    public static String SoundLevelUp = "Level_Up";
    public static String SoundMinusHabit = "Minus_Habit";
    public static String SoundPlusHabit = "Plus_Habit";
    public static String SoundReward = "Reward";
    public static String SoundTodo = "ToDo";

    @Inject
    SoundFileLoader soundFileLoader;
    private String soundTheme;

    private HashMap<String, SoundFile> loadedSoundFiles;

    public SoundManager() {
        HabiticaBaseApplication.getComponent().inject(this);

        loadedSoundFiles = new HashMap<>();
    }

    public void setSoundTheme(String soundTheme) {
        this.soundTheme = soundTheme;
    }

    public Observable<List<SoundFile>> preloadAllFiles() {
        if (soundTheme.equals("off")) {
            return Observable.empty();
        }

        ArrayList<SoundFile> soundFiles = new ArrayList<>();

        soundFiles.add(new SoundFile(soundTheme, SoundAchievementUnlocked));
        soundFiles.add(new SoundFile(soundTheme, SoundChat));
        soundFiles.add(new SoundFile(soundTheme, SoundDaily));
        soundFiles.add(new SoundFile(soundTheme, SoundDeath));
        soundFiles.add(new SoundFile(soundTheme, SoundItemDrop));
        soundFiles.add(new SoundFile(soundTheme, SoundLevelUp));
        soundFiles.add(new SoundFile(soundTheme, SoundMinusHabit));
        soundFiles.add(new SoundFile(soundTheme, SoundPlusHabit));
        soundFiles.add(new SoundFile(soundTheme, SoundReward));
        soundFiles.add(new SoundFile(soundTheme, SoundTodo));
        return soundFileLoader.download(soundFiles);
    }

    public void clearLoadedFiles() {
        loadedSoundFiles.clear();
    }

    public void loadAndPlayAudio(String type) {
        if ("off".equals(soundTheme) || soundTheme == null) {
            return;
        }

        if (loadedSoundFiles.containsKey(type)) {
            loadedSoundFiles.get(type).play();
        } else {
            ArrayList<SoundFile> soundFiles = new ArrayList<>();

            soundFiles.add(new SoundFile(soundTheme, type));
            soundFileLoader.download(soundFiles).observeOn(Schedulers.newThread()).subscribe(audioFiles1 -> {
                SoundFile file = soundFiles.get(0);

                loadedSoundFiles.put(type, file);
                file.play();

            }, RxErrorHandler.handleEmptyError());
        }
    }

}

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/SoundManager.kt
package com.habitrpg.android.habitica.helpers

import com.habitrpg.common.habitica.helpers.launchCatching
import kotlinx.coroutines.MainScope
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SoundManager
@Inject
constructor(var soundFileLoader: SoundFileLoader) {
    var soundTheme: String = SOUND_THEME_OFF

    private val loadedSoundFiles: MutableMap<String, SoundFile> = HashMap()

    fun preloadAllFiles() {
        loadedSoundFiles.clear()
        if (soundTheme == SOUND_THEME_OFF) {
            return
        }

        val soundFiles = ArrayList<SoundFile>()
        soundFiles.add(SoundFile(soundTheme, SOUND_ACHIEVEMENT_UNLOCKED))
        soundFiles.add(SoundFile(soundTheme, SOUND_CHAT))
        soundFiles.add(SoundFile(soundTheme, SOUND_DAILY))
        soundFiles.add(SoundFile(soundTheme, SOUND_DEATH))
        soundFiles.add(SoundFile(soundTheme, SOUND_ITEM_DROP))
        soundFiles.add(SoundFile(soundTheme, SOUND_LEVEL_UP))
        soundFiles.add(SoundFile(soundTheme, SOUND_MINUS_HABIT))
        soundFiles.add(SoundFile(soundTheme, SOUND_PLUS_HABIT))
        soundFiles.add(SoundFile(soundTheme, SOUND_REWARD))
        soundFiles.add(SoundFile(soundTheme, SOUND_TODO))
        MainScope().launchCatching {
            soundFileLoader.download(soundFiles)
        }
    }

    fun loadAndPlayAudio(type: String) {
        if (soundTheme == SOUND_THEME_OFF) {
            return
        }

        if (loadedSoundFiles.containsKey(type)) {
            loadedSoundFiles[type]?.play()
        } else {
            val soundFiles = ArrayList<SoundFile>()

            soundFiles.add(SoundFile(soundTheme, type))
            MainScope().launchCatching {
                val newFiles = soundFileLoader.download(soundFiles)
                val file = newFiles[0]
                loadedSoundFiles[type] = file
                file.play()
            }
        }
    }

    companion object {
        const val SOUND_ACHIEVEMENT_UNLOCKED = "Achievement_Unlocked"
        const val SOUND_CHAT = "Chat"
        const val SOUND_DAILY = "Daily"
        const val SOUND_DEATH = "Death"
        const val SOUND_ITEM_DROP = "Item_Drop"
        const val SOUND_LEVEL_UP = "Level_Up"
        const val SOUND_MINUS_HABIT = "Minus_Habit"
        const val SOUND_PLUS_HABIT = "Plus_Habit"
        const val SOUND_REWARD = "Reward"
        const val SOUND_TODO = "Todo"
        const val SOUND_THEME_OFF = "off"
    }
}
